libs/crypto/merkle-impl/src/main/kotlin/net/corda/crypto/merkle/impl/MerkleProofImpl.kt

```diff
@@ -43,6 +56,17 @@ class MerkleProofImpl(
             )
         }
 
+        // we do support and test with leaves.size == treeSize, which may not
+        // be very useful but needed not be a special case
+        if (leaves.size > treeSize) {
+            throw MerkleProofRebuildFailureException("MerkleProof has too many specified keys ${leaves.size} tree size ${treeSize}")
+        }
+        if (leaves.size < treeSize && hashes.isEmpty()) {
+            throw MerkleProofRebuildFailureException("No fill-in hashes specified to MerkleProof")
+        }
+        if (hashes.size > treeSize + leaves.size) {
```

Comment:
Would this be correct without the `+ leaves.size` ? 
```suggestion
        if (hashes.size > treeSize) {
```
I suspect the exact upper bound is somewhere around the half of the tree size.
(if you create a proof for the left(even) items on the lower level, you need the hashes of the right(odd) ones. If you add more elements to the proof, you can remove hashes. If you remove more elements from the proof, you can remove the removed element's right pair's hash since you need to add their compound hash, but essentially the number of hashes stay the same.)

Reply:
Yes, but I don't want to take the time to prove that now and this is safe.